policy/staged_policy.go

```diff
@@ -0,0 +1,129 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package policy
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+)
+
+var (
+	// DefaultStagedPolicies represents a default staged policies.
+	DefaultStagedPolicies StagedPolicies
+
+	// DefaultPoliciesList represents a default policies list.
+	DefaultPoliciesList = PoliciesList{DefaultStagedPolicies}
+)
+
+// StagedPolicies represent a list of policies at a specified version.
+type StagedPolicies struct {
+	// Cutover is when the policies take effect.
+	CutoverNanos int64
+
+	// Tombstoned determines whether the associated (rollup) metric has been tombstoned.
+	Tombstoned bool
+
+	// policies represent the list of policies.
+	policies []Policy
+}
+
+// NewStagedPolicies create a new staged policies.
+func NewStagedPolicies(cutoverNanos int64, tombstoned bool, policies []Policy) StagedPolicies {
+	return StagedPolicies{CutoverNanos: cutoverNanos, Tombstoned: tombstoned, policies: policies}
+}
+
+// Reset resets the staged policies.
+func (p *StagedPolicies) Reset() { *p = DefaultStagedPolicies }
+
+// IsDefault returns whether this is a default staged policies.
+func (p StagedPolicies) IsDefault() bool {
+	return p.CutoverNanos == 0 && !p.Tombstoned && p.hasDefaultPolicies()
+}
+
+// Policies returns the policies and whether the policies are the default policies.
+func (p StagedPolicies) Policies() ([]Policy, bool) {
+	return p.policies, p.hasDefaultPolicies()
+}
+
+// SamePolicies returns whether two staged policies have the same policy list,
+// assuming the policies are sorted in the same order.
+func (p StagedPolicies) SamePolicies(other StagedPolicies) bool {
+	currPolicies, currIsDefault := p.Policies()
+	otherPolicies, otherIsDefault := other.Policies()
+	if currIsDefault && otherIsDefault {
+		return true
+	}
+	if currIsDefault || otherIsDefault {
+		return false
+	}
+	if len(currPolicies) != len(otherPolicies) {
+		return false
+	}
+	for i := 0; i < len(currPolicies); i++ {
+		if currPolicies[i] != otherPolicies[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// String is the representation of staged policies.
+func (p StagedPolicies) String() string {
+	var buf bytes.Buffer
+	buf.WriteString(fmt.Sprintf("{cutover:%s,tombstoned:%v,policies:[", time.Unix(0, p.CutoverNanos).String(), p.Tombstoned))
+	for i := range p.policies {
+		buf.WriteString(p.policies[i].String())
+		if i < len(p.policies)-1 {
+			buf.WriteString(",")
+		}
+	}
+	buf.WriteString("]}")
+	return buf.String()
+}
+
+func (p StagedPolicies) hasDefaultPolicies() bool {
+	return len(p.policies) == 0
+}
+
+// PoliciesList is a list of staged policies.
+type PoliciesList []StagedPolicies
+
+// IsDefault determines whether this is a default policies list.
+func (l PoliciesList) IsDefault() bool {
+	return len(l) == 1 && l[0].IsDefault()
+}
+
+// WithDefaultAggregation updates the PoliciesList with default aggregation types.
+func (l PoliciesList) WithDefaultAggregation() PoliciesList {
```

Comment:
Can we rename this to `SetDefaultAggregation` and add a comment here indicating the original list will be mutated in place? Usually `With...` methods don't mutate what's passed in.